<a href="https://colab.research.google.com/github/rajitakolla/computational_methods/blob/master/In_class_exercise_08.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **The eighth in-class-exercise (20 points in total, 10/29/2020)**

The data for this exercise is from the dataset you created from assignment three. Please perform answer the following questions based on your data:

## (1) (10 points) Write a python program to extract the sentiment related terms from the corpus. You may use python package such as polyglot or external lexicon resources in the question. Rank the sentiment related terms by frequency.

In [ ]:
# Write your code here
# from polyglot.text import Text

import pandas as pd

sentiment_data = pd.read_csv("/content/sentimentAnalysis.csv")

sentiment_data.head(10)

required_df=sentiment_data[["User Names","Reviews","cleaned_data"]]

required_df.head(5)

User Names  ...                                       cleaned_data
0  morino-59166  ...  when i saw magnificent heath ledger years go e...
1   asassandroo  ...  i say im really grateful movies like one haven...
2    samanehdds  ...  im fan batman movies went see teenage son its ...
3    guillevica  ...  what master piece wondeful movie great constru...
4     ahurworth  ...  has best movie ive seen years it feel taxi dri...

[5 rows x 3 columns]

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
#remove stop words 

from nltk.corpus import stopwords
stopwords = list(stopwords.words('english'))

In [ ]:
#split the words and find the occurance of each word in the entire list

required_df["tokens"] = required_df["cleaned_data"].apply(lambda x : nltk.word_tokenize(str(x)))
required_df.head(5)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


User Names  ...                                             tokens
0  morino-59166  ...  [when, i, saw, magnificent, heath, ledger, yea...
1   asassandroo  ...  [i, say, im, really, grateful, movies, like, o...
2    samanehdds  ...  [im, fan, batman, movies, went, see, teenage, ...
3    guillevica  ...  [what, master, piece, wondeful, movie, great, ...
4     ahurworth  ...  [has, best, movie, ive, seen, years, it, feel,...

[5 rows x 4 columns]

In [ ]:
import pandas as pd
import numpy as np
def compute_values(ip):
  res1 =pd.value_counts(ip)
  if(len(res1)==0 ):
    res1 = [0]
  return res1

#calculate the occurance of each word in the sentence
new_df = required_df["tokens"].apply(lambda x : compute_values(x))

In [ ]:
new_df = new_df.fillna(0)
final_df1 = pd.DataFrame(new_df.columns, columns=["tokens"])
res = pd.DataFrame(new_df.T.sum(axis=1),columns = ["count_of_value"]).values.tolist()
final_df1["count_of_occurances"] = pd.DataFrame(res)

In [ ]:
final_df1.head(5)

tokens  count_of_occurances
0        i                109.0
1     role                  8.0
2    years                 13.0
3    heath                  7.0
4  phoenix                 42.0

In [ ]:
!pip install TextBlob

In [ ]:
from textblob  import TextBlob
#find the polarity 
final_df1["polarity"] = final_df1['tokens'].apply(lambda x: TextBlob(x).sentiment.polarity)

#filter the influencing variables 
op1 = final_df1["polarity"] != 0
op1_filter = final_df1[op1]
op1_filter.head(10)

tokens  count_of_occurances  polarity
6       amazing                 23.0  0.600000
7           new                  5.0  0.136364
8         great                 27.0  0.800000
14  magnificent                  2.0  1.000000
17   absolutely                  8.0  0.200000
31       really                 20.0  0.200000
35    beautiful                  4.0  0.850000
40  outstanding                  5.0  0.500000
41    excellent                  4.0  1.000000
48       tragic                  2.0 -0.750000

In [ ]:
#rank the words based on the occurance 
result1 = op1_filter.sort_values(by = ["count_of_occurances"],ascending = False).reset_index()


In [ ]:
result1["rank"] = pd.DataFrame(range(1,242))
result1.head(10)

index   tokens  count_of_occurances  polarity  rank
0    124     best                 40.0      1.00   1.0
1      8    great                 27.0      0.80   2.0
2      6  amazing                 23.0      0.60   3.0
3     31   really                 20.0      0.20   4.0
4    381     dark                 18.0     -0.15   5.0
5    107   mental                 15.0     -0.10   6.0
6     80     much                 13.0      0.20   7.0
7    338     good                 11.0      0.70   8.0
8    185     love                 11.0      0.50   9.0
9    442    comic                 10.0      0.25  10.0

## (2) (10 points) Compare the performance of the following tools in sentiment identification: TextBlob (https://textblob.readthedocs.io/en/dev/), VADER (https://github.com/cjhutto/vaderSentiment), TFIDF-based Support Vector Machine (SVM) (Split your data into training and testing data). Take your own annotation as the standard answers. 

Reference code: https://towardsdatascience.com/fine-grained-sentiment-analysis-in-python-part-1-2697bb111ed4

TEXTBLOB

In [ ]:
# Write your code here
required_df2 = sentiment_data.drop(["doc_id"],axis=1).reset_index()
required_df2["polarity"] = sentiment_data["cleaned_data"].apply(lambda x: TextBlob(str(x)).sentiment.polarity)
required_df2.head(5)

index    User Names  ... sentiments  polarity
0      0  morino-59166  ...   positive  0.463287
1      1   asassandroo  ...   positive  0.155357
2      2    samanehdds  ...   positive  0.322222
3      3    guillevica  ...   positive  0.750000
4      4     ahurworth  ...   positive  0.314286

[5 rows x 6 columns]

In [ ]:
required_df2['sentiment_res'] = pd.cut(required_df2['polarity'], bins=3, labels=["negative", "neutral", "positive"])
required_df2["sentiments"]=required_df2["sentiments"].fillna("negative")

In [ ]:
required_df2.head(5)

index    User Names  ...  polarity sentiment_res
0      0  morino-59166  ...  0.463287       neutral
1      1   asassandroo  ...  0.155357      negative
2      2    samanehdds  ...  0.322222       neutral
3      3    guillevica  ...  0.750000      positive
4      4     ahurworth  ...  0.314286       neutral

[5 rows x 7 columns]

VADER

In [ ]:
required_df3 = sentiment_data.drop(["doc_id"],axis=1).reset_index()
required_df3["sentiments"] = required_df3["sentiments"].fillna("negative")
required_df3.head(5)

index  ... sentiments
0      0  ...   positive
1      1  ...   positive
2      2  ...   positive
3      3  ...   positive
4      4  ...   positive

[5 rows x 5 columns]

In [ ]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


True

In [ ]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer as sa
sentiment_intensity_analyzer = sa()
required_df3["polarity_dict"] = required_df3["cleaned_data"].apply(lambda x: sentiment_intensity_analyzer.polarity_scores(str(x)))

/usr/local/lib/python3.6/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


In [ ]:
required_df3["polarity_dict"].head(10)

0    {'neg': 0.138, 'neu': 0.464, 'pos': 0.398, 'co...
1    {'neg': 0.087, 'neu': 0.528, 'pos': 0.385, 'co...
2    {'neg': 0.282, 'neu': 0.486, 'pos': 0.232, 'co...
3    {'neg': 0.0, 'neu': 0.498, 'pos': 0.502, 'comp...
4    {'neg': 0.109, 'neu': 0.582, 'pos': 0.31, 'com...
5    {'neg': 0.082, 'neu': 0.425, 'pos': 0.493, 'co...
6    {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound...
7    {'neg': 0.314, 'neu': 0.495, 'pos': 0.191, 'co...
8    {'neg': 0.133, 'neu': 0.471, 'pos': 0.396, 'co...
9    {'neg': 0.178, 'neu': 0.595, 'pos': 0.227, 'co...
Name: polarity_dict, dtype: object

In [ ]:
required_df3["polarity"] = required_df3["cleaned_data"].apply(lambda x: sentiment_intensity_analyzer.polarity_scores(str(x))["compound"])

In [ ]:
required_df3['sentiment_res'] = pd.cut(required_df3['polarity'], bins=3, labels=["negative", "neutral", "positive"])

In [ ]:
required_df3.head(5)

index    User Names  ... polarity sentiment_res
0      0  morino-59166  ...   0.9802      positive
1      1   asassandroo  ...   0.9791      positive
2      2    samanehdds  ...  -0.6002      negative
3      3    guillevica  ...   0.9442      positive
4      4     ahurworth  ...   0.8658      positive

[5 rows x 8 columns]

svm


In [67]:
from sklearn import feature_extraction, model_selection, svm, linear_model

In [75]:
required_df4 = sentiment_data[["cleaned_data","sentiments"]].reset_index()
required_df4["sentiments"] = required_df4["sentiments"].fillna("neagative")
required_df4["cleaned_data"] = required_df4["cleaned_data"].fillna("  ")

In [76]:
required_df4.head(2)

index                                       cleaned_data sentiments
0      0  when i saw magnificent heath ledger years go e...   positive
1      1  i say im really grateful movies like one haven...   positive

In [77]:
train_data, test_data = model_selection.train_test_split(required_df4, test_size = 0.5)

In [78]:
import sklearn
from sklearn.pipeline import Pipeline as pipe

pipeline = pipe([('countvector', feature_extraction.text.CountVectorizer()),('tf-idf', feature_extraction.text.TfidfTransformer()),('classifier', linear_model.SGDClassifier())])

In [96]:
vector = pipeline.fit(train_data['cleaned_data'], train_data['sentiments'])
test_data['predicted cleaned_data'] = vector.predict(test_data['cleaned_data'])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [97]:
test_data.head(5)

index  ... predicted cleaned_data
3       3  ...               positive
18     18  ...               positive
51     51  ...               positive
88     88  ...               positive
33     33  ...               positive

[5 rows x 4 columns]

Accuracy comparisions

In [98]:
from sklearn.metrics import accuracy_score

In [101]:
accuracy_score(required_df2["sentiments"],required_df2["sentiment_res"])

0.31

The accuracy score for Text Blob is 0.31

In [102]:
accuracy_score(required_df3["sentiments"],required_df3["sentiment_res"])

0.66

The accuracy score obtained by using vader is 0.66

In [100]:
accuracy_score(test_data["sentiments"].values.tolist(),test_data["predicted cleaned_data"].values.tolist())

0.76

The accuracy score obtained by using svm is 0.76